components/script/dom/bindings/structuredclone.rs

```diff
@@ -67,16 +67,62 @@ unsafe fn read_length(r: *mut JSStructuredCloneReader)
   return length as usize;
 }
 
+struct StructuredCloneWriter {
+    w: *mut JSStructuredCloneWriter,
+}
+
+impl StructuredCloneWriter {
+    unsafe fn write_slice(&self, v: &[u8]) {
+        // writers bytesl takes in a vec! and its length
+        let blob_length = v.len();
+        assert!(JS_WriteBytes(self.w, v.as_ptr() as *const raw::c_void, blob_length));
+    }
+    unsafe fn write_usize(&self, mut u: &usize) {
+        write_length(self.w, u.to_be());
+    }
+}
+
+struct StructuredCloneReader {
+    r: *mut JSStructuredCloneReader,
+}
+
+impl StructuredCloneReader {
+    unsafe fn read_bytes(&self) -> usize {
+        // reads and return length
+        let type_length = read_length(self.r);
+        return type_length;
+    }
+    unsafe fn read_str(&self) -> String {
+        // return string from vec!
+        let type_str_length = read_length(self.r);
+        let mut type_str_buffer = vec![0u8; type_str_length];
+        return String::from_utf8_unchecked(type_str_buffer);
+    }
+    unsafe fn read_slice(&self) -> Vec<u8> {
+        // returns vec! with blob/str length
+        let blob_buffer = read_length(self.r);
+        let vect = vec![0u8; blob_buffer];
+        return vect;
+    }
+    unsafe fn read_assert(&self, v: &mut [u8], l: usize) {
+        // Reads bytes; takes in vec! and length
+        let blob_length = l;
+        let vect = v.as_mut_ptr();
+        assert!(JS_ReadBytes(self.r, vect as *mut raw::c_void, blob_length));
+    }
+}
+
 unsafe fn read_blob(cx: *mut JSContext,
                     r: *mut JSStructuredCloneReader)
                     -> *mut JSObject {
-    let blob_length = read_length(r);
-    let type_str_length = read_length(r);
-    let mut blob_buffer = vec![0u8; blob_length];
-    assert!(JS_ReadBytes(r, blob_buffer.as_mut_ptr() as *mut raw::c_void, blob_length));
-    let mut type_str_buffer = vec![0u8; type_str_length];
-    assert!(JS_ReadBytes(r, type_str_buffer.as_mut_ptr() as *mut raw::c_void, type_str_length));
-    let type_str = String::from_utf8_unchecked(type_str_buffer);
+    let structured_reader = StructuredCloneReader{r: r};
+    let blob_length = structured_reader.read_bytes();
+    let type_str_length = structured_reader.read_bytes();
+    let mut blob_buffer = structured_reader.read_slice();
+    structured_reader.read_assert(&mut blob_buffer, blob_length);
+    let mut type_str_buffer = structured_reader.read_slice();
+    structured_reader.read_assert(&mut type_str_buffer, type_str_length);
+    let type_str = structured_reader.read_str();
```

Comment:
Given the changes requested previously, I would expect this code to look like:
```rust
let blob_buffer = structured_reader.read_bytes();
let type_str = structured_reader.read_str();
```
That's all.